framework/decode/deferred_operation_info_create_ray_tracing_pipelines.h

```diff
@@ -0,0 +1,108 @@
+/*
+** Copyright (c) 2021 Advanced Micro Devices, Inc. All rights reserved.
+**
+** Permission is hereby granted, free of charge, to any person obtaining a
+** copy of this software and associated documentation files (the "Software"),
+** to deal in the Software without restriction, including without limitation
+** the rights to use, copy, modify, merge, publish, distribute, sublicense,
+** and/or sell copies of the Software, and to permit persons to whom the
+** Software is furnished to do so, subject to the following conditions:
+**
+** The above copyright notice and this permission notice shall be included in
+** all copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+** FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+** FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+** DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef GFXRECON_DECODE_DEFERRED_OPERATION_INFO_CREATE_RAY_TRACING_PIPELINES_H
+#define GFXRECON_DECODE_DEFERRED_OPERATION_INFO_CREATE_RAY_TRACING_PIPELINES_H
+
+#include "generated/generated_vulkan_struct_decoders.h"
+#include "decode/vulkan_object_info.h"
+#include "decode/deferred_operation_info.h"
+
+#include "vulkan/vulkan.h"
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(decode)
+
+class DeferredOperationInfoCreateRayTracingPipelines : public DeferredOperationInfo
+{
+  public:
+    DeferredOperationInfoCreateRayTracingPipelines() :
+        DeferredOperationInfo(format::ApiCallId::ApiCall_vkCreateRayTracingPipelinesKHR)
+    {}
+
+    ~DeferredOperationInfoCreateRayTracingPipelines() {}
+
+    format::HandleId& GetDeviceId() { return device_; }
+
+    format::HandleId& GetDeferredOperationId() { return deferred_operation_; }
+
+    format::HandleId& GetPipelineCacheId() { return pipeline_cache_; }
+
+    uint32_t& GetCreateInfoCount() { return create_info_count_; }
+
+    StructPointerDecoder<Decoded_VkRayTracingPipelineCreateInfoKHR>& GetCreateInfos() { return create_infos_; }
+
+    StructPointerDecoder<Decoded_VkAllocationCallbacks>& GetAllocator() { return allocator_; }
+
+    HandlePointerDecoder<VkPipeline>& GetPipelines() { return pipelines_; }
+
+    void SetPipelineHandleInfo(std::unique_ptr<std::vector<PipelineInfo>>& pipeline_infos)
+    {
+        pipeline_handle_info_ = std::move(pipeline_infos);
+    }
+
+    std::unique_ptr<std::vector<PipelineInfo>>& GetPipelineHandleInfo() { return pipeline_handle_info_; }
```

Comment:
It's a a guarantee of lifetime. For vkCreateRayTracingPipelinesKHR, if the call is not a deferred operation, unwrapped pCreateInfos or other data will be used locally, allocated in
vkCreateRayTracingPipelinesKHR API handler and will be freed in same handler. But if the call is a deferred operation, all those data struct be just allocated in this API handler and freed at other API handler which might be in other thread when detected the deferred operation has ended.  The ownership transfer is used here.  In short, the ownership transfer will happen ( by methods like the following SetModifiedCreateInfos) when these data be allocated and freed in different API handler or the deferred operation cannot be finished in the call.

Reply:
I won't approve this PR unless you change your use of unique_ptr like I suggested above.

Reply:
I think this is harder to read than `shared_ptr` and is unsafe, and is unnecessarily confusing.  First, your code doesn't do something unsafe, but having a bare reference returned to an internal struct member means another engineer might do something unsafe.  The reference's lifetime is not tied to the object's lifetime.  Code in a totally different place is altering the contents of the member through a reference.  I've provided an example of how that could go badly at https://godbolt.org/z/Khj1eq5Ko.

Any of my suggestions above would be clearer and allow a new engineer in the code to better understand and use  this object more easily.

Reply:
I've searched all the `gfxreconstruct` code.  Unless my search was wrong, your PR will be the only addition that uses these idioms, so it will also be confusing for anyone else to use for that reason.

Reply:
ok, will change to shared_ptr.